deno_typescript/compiler_main.js

```diff
@@ -0,0 +1,332 @@
+    return fileName;
+  }
+
+  // getNewLine(): string
+  getNewLine() {
+    return "\n";
+  }
+
+  // readDirectory?(rootDir: string, extensions: ReadonlyArray<string>, excludes: ReadonlyArray<string> | undefined, includes: ReadonlyArray<string>, depth?: number): string[];
+  readDirectory() {
+    unreachable();
+  }
+
+  // resolveModuleNames?(
+  //   moduleNames: string[],
+  //   containingFile: string,
+  //   reusedNames?: string[],
+  //   redirectedReference?: ResolvedProjectReference
+  // ): (ResolvedModule | undefined)[];
+  resolveModuleNames(moduleNames, containingFile) {
+    const resolvedNames = dispatch("resolveModuleNames", {
+      moduleNames,
+      containingFile
+    });
+    const r = resolvedNames.map(resolvedFileName => {
+      const extension = getExtension(resolvedFileName);
+      return { resolvedFileName, extension };
+    });
+    return r;
+  }
+
+  // resolveTypeReferenceDirectives?(typeReferenceDirectiveNames: string[], containingFile: string, redirectedReference?: ResolvedProjectReference): (ResolvedTypeReferenceDirective | undefined)[];
+  /*
+  resolveTypeReferenceDirectives() {
+    unreachable();
+  }
+  */
+
+  // getEnvironmentVariable?(name: string): string | undefined;
+  getEnvironmentVariable() {
+    unreachable();
+  }
+
+  // createHash?(data: string): string;
+  createHash() {
+    unreachable();
+  }
+
+  // getParsedCommandLine?(fileName: string): ParsedCommandLine | undefined;
+  getParsedCommandLine() {
+    unreachable();
+  }
+}
+
+function configure(configurationText) {
+  const { config, error } = ts.parseConfigFileTextToJson(
+    "tsconfig.json",
+    configurationText
+  );
+  if (error) {
+    return { diagnostics: [error] };
+  }
+  const { options, errors } = ts.convertCompilerOptionsFromJson(
+    config.compilerOptions,
+    ""
+  );
+  return {
+    options,
+    diagnostics: errors.length ? errors : undefined
+  };
+}
+
+function dispatch(opName, obj) {
+  const s = JSON.stringify(obj);
+  const msg = encode(s);
+  const resUi8 = Deno.core.dispatch(ops[opName], msg);
+  const resStr = decodeAscii(resUi8);
+  const res = JSON.parse(resStr);
+  if (!res["ok"]) {
+    throw Error(`${opName} failed ${res["err"]}. Args: ${JSON.stringify(obj)}`);
+  }
+  return res["ok"];
+}
+
+function exit(code) {
+  dispatch("exit", { code });
+  unreachable();
+}
+
+// Maximum number of diagnostics to display.
+const MAX_ERRORS = 5;
+
+function handleDiagnostics(host, diagnostics) {
+  if (diagnostics && diagnostics.length) {
+    let rest = 0;
+    if (diagnostics.length > MAX_ERRORS) {
+      rest = diagnostics.length - MAX_ERRORS;
+      diagnostics = diagnostics.slice(0, MAX_ERRORS);
+    }
+    const msg = ts.formatDiagnosticsWithColorAndContext(diagnostics, host);
+    println(msg);
+    if (rest) {
+      println(`And ${rest} other errors.`);
+    }
+    exit(1);
+  }
+}
+
+/** Returns the TypeScript Extension enum for a given media type. */
+function getExtension(fileName) {
```

Comment:
This is interesting - on one hand I think this method won't work with extensionless files that were given content type in headers but on the other it allows to implement background download of multiple modules.

Reply:
Yes, we have two compilers now. I would love to merge them, but that is going to take further work. This is just the first step.

Reply:
How about bootstrapping? Theoretically we could make Deno to bundle code in `js/`...

Reply:
No, I think that introduces too much complexity. I don't want our build to depend on binaries.

Reply:
My plan is to slowly port js/compiler.ts to javascript and move it into deno_typescript.